### ISSUE TITLE
Inconsistent use of the "input" parameter to the Backward method in ANNs

### ISSUE THREAD

**Issue by akropp** [mlpack/mlpack#3551]
#### Issue description

It seems that there is some inconsistency in the `input` parameter of the `Backward` method on ANN layers.  The name `input` implies that it should contain the input values to the layer, and in certain cases, that is in fact what is assumed (such as in the loss functions, and in the top-level call to the network in ffn_impl.  In many of the layers, the input is actually not used at all, however, in some of the layers, it is actually assumed to be the **output** from the layer.  This is true in softmax_impl and logsoftmax_impl.  This is also true in some subset of the activation functions (but seemingly not all of them).  

The implementation of MultiLayer makes this assumption as well, passing the **output** values from each layer to its `Backward` method (as the `input` parameter) when backpropagating the errors.  Thus for the layers which are interpreting that parameter as the output, they get the expected values, and for the layers that ignore it, they don't care.  However, there are certain cases where a layer might actually need the input to implement the `Backward` method, so having this inconsistency can cause bugs (and may actually be causing some so far undetected ones).  

I propose this is either formally changed to be output (and every method that uses it is confirmed to make that assumption), or that it be correctly implemented as input (meaning the implementation of MultiLayer and of any classes that currently assume it to be output need to be changed).  I have already gone partially down this path in pr #3546, but realized there are more impacts than I thought.  I wanted to post here to get some clarity on what the desired interface/behavior is.

#### Your environment

 * version of mlpack: 4.2.1
 * operating system: any (linux)
 * compiler: any (gcc)
 * version of dependencies (Boost/Armadillo): 
 * any other environment information you think is relevant:

**Comment by rcurtin** [comment]
Thank you for pointing this out!

I spent a while digging into it, and you are correct.  I think the intention when we did the refactoring in #2777 was to pass the input to the layer, like you pointed out.  That is especially useful for some layers (`MeanPooling` and `MaxPooling` for example).  However, I see two errors: `FFN::Backward()` is passing the layer outputs to `network->Backward()` (which is `MultiLayer::Backward()`), and then all the internal calls there are off by one in the layer outputs they pass.

It is possible to rederive `LogSoftmax` and `Softmax` such that they use the input values, not the output; but, we have both available in `MultiLayer` and `FFN`.  I wonder if it is better to simply change the API of `Backward()`:

```
void Backward(const MatType& forwardInput,
              const MatType& forwardOutput,
              const MatType& gy,
              MatType& delta);
```

I'd be curious what @zoq and @Aakash-kaushik and others thought before we moved on that.  It wouldn't be a high-effort refactoring and I don't see any downsides (except the marginally increased complexity added by another parameter), but maybe I overlooked something.  Also related: #3470 and @mrdaybird's related observations there.

**Comment by akropp** [comment]
This would certainly be the most flexible solution, as some layers need the input and some benefit from having the output (in efficiency).  At the very least, I think it should be changed to the input.  An alternative (though I guess maybe less memory-efficient) is done already in certain places, is where any layers that need their outputs in the backwards/gradient methods just remember it during the forward pass. It already has the contract that backward is not called until Forward has been called.

I am happy to do the re-factoring if you want.  I have already done some of it in the branch mentioned above (when I started, I didn't realize the extent of the issue).  I can pull the relevant changes over to a new branch explicitly for this issue.

**Comment by rcurtin** [comment]
I chatted with @zoq, he agrees that adding an extra parameter is a reasonable thing to do.  If you're up for the refactoring I would really appreciate that, and I can review once it's ready.  Up to you whether you want to do both as part of #3550 or make a new PR---the changes are orthogonal enough that while reviewing I should be able to keep them separated in my head.  Thanks again for pointing this out. :+1:

**Comment by akropp** [comment]
I guess it would be simpler to do iit all together, since I've already started down that path.  Just wasn't sure if you prefer having separate changes like this merged separately.

**Comment by akropp** [comment]
Committed (most of) the changes needed to the branch for #3547.  The one remaining bit (that I know of) is that I believe most of the activation functions are written assuming that Deriv() is passed the output from the layer, but some (e.g. InvQuadFunction) seem to be written assuming the input will be passed to Deriv().  I can try to go through to identify which is which, but then what do we do about it?  

My thought was to perhaps have a FunctionTraits<> template class which indicates whether Deriv expects the output or the input.  That was whichever one is more efficient to implement could be used.  BaseLayer could have two implementations of Backward based on the indicator.  Either that, or have two differently named methods on the activation functions depending on which one is expected.

**Comment by rcurtin** [comment]
Yeah, sometimes the derivative is actually easier to express in terms of the input.  If `Backward()` already needs both the inputs and the outputs available, I feel like there's a good argument for making exactly the same change to `Deriv()`, and supplying both the input and output.  I took a quick look through and saw that `Softplus` also wants the input, not the output.

The `FunctionTraits` idea could work, although it wouldn't cover the case where it happens to be maximally convenient to express the derivative in terms of the output *and* the input.  (For example take the random function I just made up `y = exp(x^2)`, where the derivative `dy = exp(y) 2x`.)  But, happy to put off considering that situation until it's actually encountered.

@mrdaybird's work in https://github.com/mlpack/mlpack/pull/3478#issuecomment-1584899616 is directly relevant for this second issue (it is exactly the same issue).  In fact the solution proposed there is basically equivalent to using `FunctionTraits`, although in this case instead of two different classes, we can keep the code all in `BaseLayer` and just branch according to the compile-time property of whether or not we want the input or output.

**Comment by akropp** [comment]
Yeh, I guess the cleanest/most flexible solution might be to pass both the input and output to Deriv, although that feels a little strange for some reason. ;). If this were a general function abstraction, I'm not sure I'd like that, but considering it is only used here, it's probably ok.

If we went for the Deriv from either input or output, I guess it could also be something like you just implement one of DerivFromX or DerivFromY.

I'm happy to change it to whichever you think makes the most sense here.

**Comment by akropp** [comment]
What if I make activation functions support one (and only one) of the following:
DerivFromInput
DerivFromOutput
DerivFromBoth

The implementer could choose to implement whichever one makes the most sense.  The name makes it clear what you are implementing.  BaseLayer could then have three versions of Backward, depending on which one of these is implemented.

Or... I could just make it take both all the time.

**Comment by rcurtin** [comment]
Up to you, I don't have a particularly strong opinion here.  It's probably a nice thing to do to rename the function to indicate whether it's input or output. :+1:

**Comment by rcurtin** [comment]
Fixed in #3547. :+1:
